File: sdk-java/src/test/java/io/superdurable/dex/integ/InternalChannelWorkflow.java
/*
 * Portions of this file are derived from indeedeng/iwf-java-sdk.
 * Those portions are licensed under the Apache License, Version 2.0.
 * See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
 *
 * Modifications Copyright (c) 2026 Super Durable, Inc.
 *
 * Modifications are licensed under the Super Durable Source License 1.0.
 * Third-Party Materials remain under the Apache License, Version 2.0.
 * See LICENSE and LEGACY_NOTICES.md.
 */

package io.superdurable.dex.integ;

import io.superdurable.dex.Channel;
import io.superdurable.dex.ChannelMap;
import io.superdurable.dex.ConditionCombination;
import io.superdurable.dex.Context;
import io.superdurable.dex.Flow;
import io.superdurable.dex.PersistenceSchema;
import io.superdurable.dex.Step;
import io.superdurable.dex.StepList;
import io.superdurable.dex.StepDecision;
import io.superdurable.dex.StepMovement;
import io.superdurable.dex.Wait;

final class InternalChannelWorkflow implements Flow<Integer> {
    private final Channel<Integer> firstChannel =
            Channel.define("test-inter-state-channel-1", Integer.class);
    private final Channel<Integer> secondChannel =
            Channel.define("test-inter-state-channel-2", Integer.class);
    private final ChannelMap<Integer> channelMap =
            ChannelMap.define("test-inter-state-channel-map", Integer.class);
    private final ForkStep start = new ForkStep();
    private final ConsumeStep consumer = new ConsumeStep();
    private final PublishStep publisher = new PublishStep();

    @Override
    public StepList<Integer> getSteps() {
        return StepList.startStep(start).otherSteps(consumer, publisher);
    }

    @Override
    public PersistenceSchema getPersistenceSchema() {
        return PersistenceSchema.of(firstChannel, secondChannel, channelMap);
    }

    final class ForkStep implements Step<Integer> {
        @Override
        public Class<Integer> getInputType() {
            return Integer.class;
        }

        @Override
        public StepDecision execute(final Context context, final Integer input) {
        return StepDecision.goToMany(
                    StepMovement.of(ConsumeStep.class, input),
                    StepMovement.of(PublishStep.class, input));
        }
    }

    final class ConsumeStep implements Step<Integer> {
        @Override
        public Class<Integer> getInputType() {
            return Integer.class;
        }

        @Override
        public Wait waitFor(final Context context, final Integer input) {
            return Wait.anyCombinationOf(
                    ConditionCombination.of(
                            firstChannel.forOne("first"),
                            channelMap.forOne("one", "mapped")),
                    ConditionCombination.of(secondChannel.forOne("second")));
        }

        @Override
        public StepDecision execute(final Context context, final Integer input) {
            if (!secondChannel.getConditionResults(context).isEmpty()) {
                throw new IllegalStateException("second channel should still be waiting");
            }
            final Integer firstValue = firstChannel.getConditionResults(context).get(0);
            final Integer mappedValue = channelMap.getConditionResults(context, "one").get(0);
            if (mappedValue.intValue() != 3) {
                throw new IllegalStateException("mapped channel returned " + mappedValue);
            }
            return StepDecision.gracefulComplete(input + firstValue);
        }
    }

    final class PublishStep implements Step<Integer> {
        @Override
        public Class<Integer> getInputType() {
            return Integer.class;
        }

        @Override
        public StepDecision execute(final Context context, final Integer input) {
            firstChannel.publish(context, 2);
            channelMap.publish(context, "one", 3);
            return StepDecision.deadEnd();
        }
    }
}

File: sdk-rust/crates/dex-sdk/tests/integ/internal_channel_workflow.rs
// Portions of this file are derived from indeedeng/iwf-java-sdk.
// Those portions are licensed under the Apache License, Version 2.0.
// See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
//
// Modifications Copyright (c) 2026 Super Durable, Inc.
//
// Modifications are licensed under the Super Durable Source License 1.0.
// Third-Party Materials remain under the Apache License, Version 2.0.
// See LICENSE and LEGACY_NOTICES.md.

use std::sync::LazyLock;

use dex_sdk::{
    Channel, ChannelMap, ConditionCombination, Context, Flow, HandlerError, HandlerResult,
    PersistenceSchema, Step, StepDecision, StepList, StepMovement, Wait,
};

static FIRST_CHANNEL: LazyLock<Channel<i32>> =
    LazyLock::new(|| Channel::new("test-inter-state-channel-1"));
static SECOND_CHANNEL: LazyLock<Channel<i32>> =
    LazyLock::new(|| Channel::new("test-inter-state-channel-2"));

pub(crate) struct InternalChannelWorkflow {
    first_channel: Channel<i32>,
    second_channel: Channel<i32>,
    channel_map: ChannelMap<i32>,
    start: ForkStep,
    consumer: ConsumeStep,
    publisher: PublishStep,
}

impl InternalChannelWorkflow {
    pub(crate) fn new() -> Self {
        let first_channel = FIRST_CHANNEL.clone();
        let second_channel = SECOND_CHANNEL.clone();
        let channel_map = ChannelMap::new("test-inter-state-channel-map");
        Self {
            start: ForkStep,
            consumer: ConsumeStep {
                first_channel: first_channel.clone(),
                second_channel: second_channel.clone(),
                channel_map: channel_map.clone(),
            },
            publisher: PublishStep {
                first_channel: first_channel.clone(),
                channel_map: channel_map.clone(),
            },
            first_channel,
            second_channel,
            channel_map,
        }
    }
}

impl Flow for InternalChannelWorkflow {
    type StartInput = i32;

    fn steps(&self) -> StepList<'_, Self::StartInput> {
        StepList::start(&self.start)
            .and(&self.consumer)
            .and(&self.publisher)
    }

    fn persistence(&self) -> PersistenceSchema {
        PersistenceSchema::new()
            .channel(&self.first_channel)
            .channel(&self.second_channel)
            .channel_map(&self.channel_map)
    }
}

struct ForkStep;

impl Step for ForkStep {
    type Input = i32;

    fn execute(&self, _context: &mut Context, input: i32) -> HandlerResult<StepDecision> {
        Ok(StepDecision::go_to_many([
            StepMovement::to(
                &ConsumeStep {
                    first_channel: FIRST_CHANNEL.clone(),
                    second_channel: SECOND_CHANNEL.clone(),
                    channel_map: ChannelMap::new("test-inter-state-channel-map"),
                },
                input,
            ),
            StepMovement::to(
                &PublishStep {
                    first_channel: FIRST_CHANNEL.clone(),
                    channel_map: ChannelMap::new("test-inter-state-channel-map"),
                },
                input,
            ),
        ]))
    }
}

struct ConsumeStep {
    first_channel: Channel<i32>,
    second_channel: Channel<i32>,
    channel_map: ChannelMap<i32>,
}

impl Step for ConsumeStep {
    type Input = i32;

    fn wait_for(&self, _context: &mut Context, _input: i32) -> HandlerResult<Wait> {
        Ok(Wait::any_combination_of([
            ConditionCombination::all_of([
                self.first_channel.for_one().with_id("first"),
                self.channel_map.for_one("one").with_id("mapped"),
            ]),
            ConditionCombination::all_of([self.second_channel.for_one().with_id("second")]),
        ]))
    }

    fn execute(&self, context: &mut Context, input: i32) -> HandlerResult<StepDecision> {
        if !self.second_channel.condition_results(context)?.is_empty() {
            return Err(HandlerError::new(
                "InternalChannelFailure",
                "second channel should still be waiting",
            ));
        }
        let first = self.first_channel.condition_results(context)?[0];
        let mapped = self.channel_map.condition_results(context, "one")?[0];
        if mapped != 3 {
            return Err(HandlerError::new(
                "InternalChannelFailure",
                format!("mapped channel returned {mapped}"),
            ));
        }
        Ok(StepDecision::graceful_complete(input + first))
    }
}

struct PublishStep {
    first_channel: Channel<i32>,
    channel_map: ChannelMap<i32>,
}

impl Step for PublishStep {
    type Input = i32;

    fn execute(&self, context: &mut Context, _input: i32) -> HandlerResult<StepDecision> {
        self.first_channel.publish(context, 2)?;
        self.channel_map.publish(context, "one", 3)?;
        Ok(StepDecision::dead_end())
    }
}
